docs/decisions/0035-embedding-abstraction.md

```diff
@@ -0,0 +1,257 @@
+
+### Option 1 [Current] - Generic embeddings generation interface and specialized interfaces for different types of data
+
+The current abstraction of generating embeddings only allows for generation from text and image data.\
+We return raw data in the form of `IList<ReadOnlyMemory<TEmbedding>>` instead of a specialized data type like `IList<EmbeddingContent>`.
+
+In this option, we cannot parameterize the query or return metadata.
+
+```csharp
+public interface IEmbeddingGenerationService<TValue, TEmbedding> : IAIService where TEmbedding : unmanaged
+{
+    Task<IList<ReadOnlyMemory<TEmbedding>>> GenerateEmbeddingsAsync(
+        IList<TValue> data,
+        Kernel? kernel = null,
+        CancellationToken cancellationToken = default);
+}
+
+public interface ITextEmbeddingGenerationService : IEmbeddingGenerationService<string, float> { }
+
+public interface IImageEmbeddingGenerationService : IEmbeddingGenerationService<ImageContent, float> { }
+```
+
+Pros:
+- Allows for generating embeddings from different types of data.
+- Generic interface.
+- Interface segregation.
+
+Cons:
+- Cannot parameterize the query.
+- Cannot return metadata.
+- We lost precision.
+
+### Option 2 [Proposed] - Specialized embeddings generation interfaces for different types of data with metadata and query parameterization without a generic interface
+
+This option allows you to generate embeds from different data types, such as text and images.\
+And allows you to parameterize the query and return metadata.\
+We can also develop interfaces with other returned types like `EmbeddingContent<int>`, `EmbeddingContent<double>`, etc.
+
+```csharp
+public class EmbeddingContent<TEmbedding> : KernelContent where TEmbedding : unmanaged
+{
+    public EmbeddingContent(
+        IReadOnlyList<ReadOnlyMemory<TEmbedding>> data,
+        string? modelId = null,
+        object? innerContent = null,
+        IReadOnlyDictionary<string, object?>? metadata = null)
+        : base(innerContent, modelId, metadata)
+    {
+        this.Data = data;
+    }
+
+    public IReadOnlyList<ReadOnlyMemory<TEmbedding>> Data { get; set; }
+}
+
+public interface ITextEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<string> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+
+public interface IImageEmbeddingGenerationService : IAIService
+{
+    Task<IReadOnlyList<EmbeddingContent<double>>> GenerateEmbeddingsAsync(
+        IList<ImageContent> data,
+        Kernel? kernel = null,
+        PromptExecutionSettings? executionSettings = null,
+        CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- Allows for generating embeddings from different types of data.
+- Allows for parameterizing the query.
+- Allows for returning metadata.
+- Interface segregation.
+
+Cons:
+- No generic interface.
```

Comment:
That's not always a disadvantage, but in this case, in theory, if we keep generic `IEmbeddingGenerationService<TValue, TEmbedding>` as it is today, but update `GenerateEmbeddingsAsync` method to return `IReadOnlyList<EmbeddingContent<TEmbedding>>` and add `PromptExecutionSettings` as optional parameter, we should resolve the disadvantage provided in this list and cover all required cases, is that correct? I think we should add this approach as another option in this ADR.

Then, if specific AI connector supports embeddings for texts and images, it could implement `IEmbeddingGenerationService` interface multiple times with different data types:
![image](https://github.com/microsoft/semantic-kernel/assets/13853051/621fe5fe-7ed1-4093-af4b-cf44e87b052a)

Reply:
@dmytrostruk 
Personally, I prefer the current option No 4 but I have also added a generic version as No 5.

Reply:
@dmytrostruk And do we really need generic EmbeddingContent class? Do you see any cases where  `double` type can be insufficient?

Reply:
https://github.com/microsoft/semantic-kernel/pull/5279#discussion_r1511330971 this has confused me a bit, maybe option 2 is best after all?